### DISCUSSION TITLE
How do I change the output directory in vercel.json

### DISCUSSION THREAD

**Discussion by jakobrosenberg** [vercel/vercel#5081] [Help] [answered]
I need to use the build option in vercel.json, but I can't find anything about changing the output directory in the documentation.

It can be changed in the GUI, so I expect it can also be changed in the config file.

**Comment by paulogdm** [comment]
Hello @jakobrosenberg ,

Can you share your "vercel.json" here?

**Comment by jakobrosenberg** [comment]
```json
{
  "version": 2,
  "builds": [
    { "src": "dist/**", "use": "@vercel/static"},
    { "src": "serverless/*.js", "use": "@vercel/node" }
  ],
  "routes": [
    {
      "handle": "filesystem"
    },
    {
      "src": "/.*",
      "dest": "/serverless/test.js"
    }
  ]
}
```

**Reply by paulogdm** [reply]
Are you sure you don't want to move to a [Zero Config](https://vercel.com/blog/zero-config) setup? Here is what you need to do so:

1. Remove your "vercel.json";
2. Move "serverless/**" to "api/";
3. Edit your project's configuration to make "dist/" the output folder.

Would that work for you?

**Reply by paulogdm** [reply]
If, not, here is what I recommend:
```
{
  "version": 2,
  "builds": [
    { "src": "package.json", "use": "@vercel/static-build", "config": {"distDir": "dist"} },
    { "src": "serverless/*.js", "use": "@vercel/node" }
  ],
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/serverless/test.js"
    }
  ]
}
```
EDIT: Then you need a build command on your package.json with the following command `mv dist/ .` or something similar. Again, I highly recommend that you use Zero Config if you can so you can have a better experience overall.

**Comment by jakobrosenberg** [comment]
Thank you, the distDir is what I was looking for.

Is `rewrites` preferred over `routes`?

EDIT: tried `rewrites`, but it consumes all requests, so the `dist` content is no longer accessible.

EDIT 2: It builds the `dist` folder correctly, but it doesn't see `dist` nor is able to serve `dist`'s contents.

Current `vercel.json`
```json
{
  "version": 2,
  "builds": [
    {
      "src": "package.json",
      "use": "@vercel/static-build",
      "config": { "distDir": "dist" }
    },
    {
      "src": "_roxi/.roxi/*.js",
      "use": "@vercel/node",
      "config": { "includeFiles": "dist/**" }
    }
  ]
}
```

![image](https://user-images.githubusercontent.com/4153004/91045502-45542480-e617-11ea-9434-b5abd51b8f4f.png)

**Reply by paulogdm** [reply]
My apologies. I edited my answer.

**Reply by paulogdm** [reply]
Yes, rewrites are way better. Please refer to: https://vercel.com/docs/configuration#project/rewrites

**Comment by jakobrosenberg** [comment]
@paulogdm after some experimenting, it seems that nested dirs in `dist` aren't available. Only files in the root of `dist` are accessible.

**Reply by paulogdm** [reply]
We would need a reproduction for that. Please give us the full context so we can assist you further.

**Comment by jakobrosenberg** [comment]
> EDIT: Then you need a build command on your package.json with the following command mv dist/ . or something similar. Again, I highly recommend that you use Zero Config if you can so you can have a better experience overall.

Zero config is unfortunately not an option. I have developers deploying to Vercel, Netlify, Github pages, Firebase, Electron, etc.. I worry that If I start adding service specific files/folders in the root of the workspace it will become a mess and cause conflicts. 
_(This isn't an issue exclusive to deployments, bundlers like Vite and Snowpack also have requirements about file structure. Eg. Vite is hardcoded to use `public` as a source folder for assets and `dist` as an output folder.)_

I won't be able to move the dist content to the root either. This will expose the entire project as well as mix dist with source code which would be unmaintainable.

Is there really no way to serve a static site from a specific folder? I mean it's already working, partly.
```
dist/
  hello.html      //accessible on  `/hello.html`
  folder/
    hello.html    //not accessible on `/folder/hello.html`
```

**Comment by jakobrosenberg** [comment]
I've created a reproduction here https://github.com/jakobrosenberg/vercel-dev-bug/tree/master

```
/                     // works
/hello.html           // works
/folder               // works
/folder/hello.html    // works when deployed, but is 404 with "vercel dev"
```

As shown, this may work after all. At least in production.

**Comment by jakobrosenberg** [comment]
I've hit a roadblock. "includeFiles" has no effect when used in "builds" and if I use it with a function I get
```
Error! The `functions` property cannot be used in conjunction with the `builds` property. Please remove one of them.
```

It seems it's impossible for the function to access the content built by `static-build`.

**Reply by ThatGuySam** [reply]
Same. 

I always end up having to commit files to git or make an HTTP request to itself. 

When you deploy with the Vercel CLI it will included the built artifacts so theoretically you could probably get it to work via Github Action if you could get the Vercel CLI working within the Action.

**Comment by Byron2016** [comment]
I still have the problem, I openned a new insue [How do I change the output directory in vercel.json #9818](https://github.com/vercel/vercel/issues/9818) maybe @jakobrosenberg or @paulogdm could you give me a hand with it 😩?

**Reply by paulogdm** [reply]
You can use the UI now: https://vercel.com/docs/concepts/deployments/configure-a-build#build-and-development-settings

**Comment by paulogdm** [comment, marked as answer]
You can now use the UI to change the Output Directory of your build command.
Check [our docs](https://vercel.com/docs/concepts/deployments/configure-a-build#build-and-development-settings) for more information.
